### PR TITLE
8297041: Remove the last remnants of sjavac

### PR DESCRIPTION
**Pull request by magicus**
The sjavac ("smart javac") was an ambitious project. It should parallelize java compilation, create a background daemon process that kept the JVM "hot" with the JITted javac code, define a public api so only noticeable changes in a class caused further dependency compilations, etc etc.

Some of this never came to fruition. Other were implemented differently, as the `depend` plugin. The only thing we're currently using (and have been for the last few years) is the "server" functionality, that is, the ability to keep a single JVM process alive, and reuse the JITted code.

This code does not belong in the jdk.compiler module. It is a buildtool, pure and simple. Let's move it to it's proper place.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297041](https://bugs.openjdk.org/browse/JDK-8297041): Remove the last remnants of sjavac

### Reviewers
 * [Christian Stein](https://openjdk.org/census#cstein) (@sormuras - Committer) ⚠️ Review applies to [2124c86b](https://git.openjdk.org/jdk/pull/11185/files/2124c86b625d6fa39b4260b93d58d40aecbbcfa8)
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [2124c86b](https://git.openjdk.org/jdk/pull/11185/files/2124c86b625d6fa39b4260b93d58d40aecbbcfa8)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [2124c86b](https://git.openjdk.org/jdk/pull/11185/files/2124c86b625d6fa39b4260b93d58d40aecbbcfa8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11185/head:pull/11185` \
`$ git checkout pull/11185`

Update a local copy of the PR: \
`$ git checkout pull/11185` \
`$ git pull https://git.openjdk.org/jdk pull/11185/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11185`

View PR using the GUI difftool: \
`$ git pr show -t 11185`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11185.diff">https://git.openjdk.org/jdk/pull/11185.diff</a>

</details>

### CODE REVIEW

**Comment by magicus** [comment]
Some more detailed explanations of the changes I've made in this patch.

* I have stripped down the original sjavac code to retain only the "javac server" part, and associated machinery. In some cases it can probably be possible to cut out even more baggage, but I had to draw the line somewhere.
* The remaining sjavac code has moved to `make/langtools/tools`, together with the other langtools buildtools.
* All sjavac jtreg tests in langtools has been removed.
* A few non-sjavac-related tests (for AppCDS) references sjavac. My interpretation is that these classes were just picked arbitrarily as a good representation, so I have modified these tests to pick other, arbitrary classes. (I'd like to have some input from AppCDS folks that I did this correctly)
* I have verified that it is still possible to build with, and without, the javac server, and that there has been no build performance regressions.
* The original implementation used a class internal to the `jdk.compiler` module to actually kick off the javac compilation. This was no longer an option when moving out of the module, but fortunately there is actually an exported class that does the same, `com.sun.tools.javac.Main`. This is however deprecated, and I realize it is not ideal to introduce a dependency to a deprecated class (but see below).

On a longer term, the current javac server tool is under scrutiny, and it will probably be rewritten into a more general "portal" to run other langtools binaries such as javadoc. It will also be transformed into using the ToolProvider API, which will solve the problem of calling the deprecated method. @sormuras is working on this.

**Review by sormuras — APPROVED** [review]
LGTM

**Comment by sormuras** [comment]
> Or is the replacement so imminent that it's not worth the effort?

This. I aim to replace `javacserver` with a well-tested `javatoolportal` buildtool in the coming month. An almost working POC is already available - but forcing it to go live with the JDK 20 train would not be good and replacing a minimal `javacserver` buildtool is an easier/lighter task.

**Review by erikj79 — APPROVED** [review]

**Comment by magicus** [comment]
@erikj79 I'd also like to add that, at this point, the buildtool is not very complex. Right now it just listens to a port, and kicks off a javac compilation for each incoming call. It looks more complex since we've come here the long way around from a more complicated tool, and I did not spend too much time simplifying the code.

File: make/langtools/tools/javacserver/comp/SjavacImpl.java [hunk 1]
```diff
@@ -23,57 +23,54 @@
...
-import javax.tools.JavaFileObject;
+import com.sun.tools.javac.Main;
+import javacserver.Util;
+import javacserver.Log;
+import javacserver.Result;
+import javacserver.options.Option;
+import javacserver.server.Sjavac;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.stream.Stream;
```
Comment (jonathan-gibbons) [inline make/langtools/tools/javacserver/comp/SjavacImpl.java:36]: This is a build tool, so you get to set the guidelines, but generally, this is an unusual order for imports

File: make/langtools/tools/javacserver/comp/SjavacImpl.java [hunk 2]
```diff
@@ -23,57 +23,54 @@
...
+ * The sjavac implementation that interacts with javac and performs the actual
+ * compilation.
  *
  *  <p><b>This is NOT part of any supported API.
  *  If you write code that depends on this, you do so at your own risk.
  *  This code and its internal interfaces are subject to change or
  *  deletion without notice.</b>
  */
-public class SmartWriter extends Writer {
+public class SjavacImpl implements Sjavac {
```
Comment (jonathan-gibbons) [inline make/langtools/tools/javacserver/comp/SjavacImpl.java:47]: ... so it looks like there are still some vestiges of `Sjavac` ...

File: test/langtools/ProblemList.txt [hunk 3]
```diff
@@ -57,18 +57,6 @@ tools/javac/modules/SourceInSymlinkTest.java
...
-###########################################################################
-#
-# sjavac
-
-tools/sjavac/IncCompileFullyQualifiedRef.java                                   8152055    generic-all    Requires dependency code to deal with in-method dependencies.
-tools/sjavac/IncCompileWithChanges.java                                         8152055    generic-all    Requires dependency code to deal with in-method dependencies.
-
-tools/sjavac/ApiExtraction.java                                                 8158002    generic-all    Requires investigation
-tools/sjavac/IgnoreSymbolFile.java                                              8158002    generic-all    Requires investigation
-tools/sjavac/ClasspathDependencies.java                                         8158002    generic-all    Requires investigation
```
Comment (jonathan-gibbons) [inline test/langtools/ProblemList.txt:69]: I presume these issues, 8152055, 8158002, are still open?    They should either be closed or assigned the the build team.

**Review by jonathan-gibbons — APPROVED** [review]
Nice cleanup.

File: make/langtools/tools/javacserver/comp/SjavacImpl.java (hunk 1 above)
Comment (magicus) [reply make/langtools/tools/javacserver/comp/SjavacImpl.java:36]: Haha! Calling this "unusual" is the most diplomatic expression I've heard today. :-) This is just bonkers. I used a combination of "organize imports" on IntelliJ (without setting up specific rules for the project) and a brute search-and-replace. I'll fix.

File: test/langtools/ProblemList.txt (hunk 3 above)
Comment (magicus) [reply test/langtools/ProblemList.txt:69]: No, I've closed them already, with a reference to this issue.

Comment (magicus) [reply test/langtools/ProblemList.txt:69]: To be clear: they referred to the test code that is removed as part of this PR, so closing them is the only reasonable thing to do.

**Pull request closed without merging**